sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs

```diff
@@ -113,21 +66,49 @@ protected override Request CreateRequest(string[] scopes)
             return request;
         }
 
+        public async override ValueTask<AccessToken> AuthenticateAsync(bool async, TokenRequestContext context, CancellationToken cancellationToken)
+        {
+            try
+            {
+                return await base.AuthenticateAsync(async, context, cancellationToken).ConfigureAwait(false);
+            }
+            catch (RequestFailedException e)
+            {
+                if (e.Status == 0)
+                {
+                    throw new CredentialUnavailableException(CannotConnectError, e);
+                }
+
+                throw;
+            }
+            catch (AggregateException e)
```

Comment:
I'm not sure I understand your question. Are you saying I should check the inner exceptions of the AggregateException to see if they are RequestFailedException? In this case I don't think it matters much because I only alter the exception message slightly so it makes more sense with the AggregateException message. In the end we're throwing the same exception type.

Reply:
With retries enabled I wonder whether we would ever get RequestFailedException thrown from the pipeline? If we want a consistent message we need to look at AggregateException inner exceptions.

Reply:
I plan to dig into error messaging and tracing over the next preview to make sure we haven't regressed the experience with this change. For for this preview I think this is sufficient, unless you're concerned about this plan.

Reply:
Sounds good!